MOpt implementation

icchy opened the pull request:
# MOpt implementation

## Type of PR
<!-- This project uses issue-driven development, so please take the appropriate action for each PR type. -->

- [x] Changes related to the roadmap (e.g., TODO.md) (type: A) -> Create an issue corresponding to the PR in advance, and refer to this PR in the issue.
- Changes that are not related to the roadmap
    - [ ] Change with multiple possible solutions to the issue (type: B-1) -> Create an issue corresponding to the PR in advance and refer to this PR in the issue.
    - [ ] Change with a single solution (type: B-2) -> There is no need to create an issue corresponding to the PR in advance. Please discuss it in this PR.

## Related Issue
<!-- If this PR is a PR type A/B-1, please provide a link to the corresponding issue. -->
<!-- If this PR is a PR type B-1, please write "N/A" -->

## Importance of PR
<!-- Please describe the importance of the PR in terms of the following aspects. -->

- Importance of the issue
    - [x] Large (based on several days to weeks of discussion and verification, e.g., this issue is a blocking issue for other issues on the roadmap, etc.)
    - [ ] Medium (based on a few hours to a day of discussion and verification, e.g., this issue is a blocking issue for another minor issue)
    - [ ] Small (apparent changes such as build error)
- Complexity of the solution (code, tests, etc.)
    - [x] Large (requires several days to several weeks of review)
    - [ ] Medium (requires several hours to a day of review)
    - [ ] Small (trivial changes, such as build error)

## PR Overview
<!-- Please provide a summary of this PR. -->
<!-- If this PR is a PR type A/B-1, this PR will be considered as an item in the checklist for the related issue. Please provide a link to the issue comment that contains the checklist. -->
<!-- If this PR is a PR type B-2, unnecessary to reference the issue. Please provide a summary. -->

This PR implements MOpt algorithm proposed by [lyu, USENIX'19](https://www.usenix.org/conference/usenixsecurity19/presentation/lyu).

Tests and Docs for MOpt are also included in this PR.

## Concerns (Optional)
<!-- If you have any concerns, please describe them clearly by filling in the relevant checklist items below. If there is anything else you would like to share with the reviewer, please include it. -->

- [ ] Performance
- [x] Source Code Quality

---

> The PR author should fill in the following checklist when submitting this PR.

#### Optional Entries
- [ ] If this PR is a PR type A/B-1, there is a cross-link between this PR and the related issue.

#### Mandatory Entries
- [ ] The PR title is a summary of the changes.
- [ ] Completed each required field of the PR.

---
> The PR author should fill out the following checklist in the comments to confirm that this PR is ready to be merged

- [ ] CI is green or confirmed test run results.
- [ ] All change suggestions from reviewers have been resolved (fixed or foregone).

---
> The maintainer of this repository will set up a reviewer for each PR.
> PR reviewers should review this PR in terms of the checklist below before moving on to a detailed code review. Please comment on their initial response by filling in the checklist below.

#### Optional Entries
- [ ] The reviewer assigned more reviewers if needed.
- [ ] The reviewer noted that it is necessary to break out some of the changes in this PR into other PRs if needed.
- [ ] The reviewer noted that the initial response is insufficient if needed.

#### Mandatory Entries
- [x] The title of this PR summarizes the changes made by this PR properly.
- [x] The target branch of this PR is as intended.
- [x] The reviewer understands the issues in this PR.
- [x] The reviewer plans to review with an appropriate workload based on the importance of this PR.

---
> When the PR reviewer concludes that this PR is ready to be merged, please fill in the checklist below by posting it in the comment. If there is more than one reviewer, please do this on your own.

#### Optional Entries
- [ ] The reviewer noted that if you believe that new tests are needed to evaluate this PR, they have been noted.
- [ ] If minor refactorings are not mentioned in the PR, I understand the intent.
- [ ] If this PR is a PR type A/B-1, we have agreed on this PR's design, direction, and granularity in the related issue.

#### Mandatory Entries
- [ ] The reviewer understands how this PR addresses the issue and the specific changes.
- [ ] This PR uses the best possible issue resolution that the reviewer can think of.
- [ ] This PR is ready to be merged.

icchy committed:
add optimizer class

potetisensei committed:
fix errors in store

potetisensei committed:
format optimizer.hpp

potetisensei committed:
refactor optimizer

potetisensei committed:
redesign Store

potetisensei committed:
add StoreKeys

potetisensei committed:
include AFLDictData in optimizer::keys

potetisensei committed:
add reference in optimizer::keys::Extras, AutExtras

potetisensei committed:
add MutopOptimizer and transform AFLHavocCaseDistrib into its derived class

potetisensei committed:
remove MutopOptimizer

potetisensei committed:
remove unnecessary classes

potetisensei committed:
add member variable mutop_optimizer in AFLState

potetisensei committed:
remove old include headers

potetisensei committed:
move Store key definition keys.cpp -> keys.hpp

potetisensei committed:
shorten namespace of AFLHavocCaseDistrib

potetisensei committed:
use NullableRef for optimizer::keys::Extras and AutoExtras

potetisensei committed:
fix wrong include

potetisensei committed:
fix compilation errors related to Mutator::Havoc

potetisensei committed:
support new Havoc in IJON and PyFuzzer

potetisensei committed:
fix compilation errors in IJON

potetisensei committed:
fix compilation error in AFLState

potetisensei committed:
refactor functions that register fuzzers

potetisensei committed:
modify tests according to changes

icchy committed:
add skeleton for MOpt

icchy committed:
add opt value SelectedCaseHistogram

icchy committed:
add optimizer for PSO

icchy committed:
add splice_cycles limit utils

icchy committed:
implement MOptHavocCaseDistrib

icchy committed:
use fuzzuf::utils::random::Random

icchy committed:
add MOpt optimizer

icchy committed:
update selected_case_histogram

icchy committed:
fix MOptHavocCaseDistrib initializer

icchy committed:
add mopt-specific hierarflow

icchy committed:
add MOpt sources

icchy committed:
fix error partially

Commit:
Fix cyclic include

icchy committed:
fix error partially

icchy committed:
fix option argument

- AFLStateTemplate<Testcase> -> State

icchy committed:
[WIP] fix compile error

icchy committed:
fix compile error

icchy committed:
Merge branch 'master' into feature/mopt

icchy committed:
fix build error due to master changes

icchy committed:
fix format

icchy committed:
add CLI for MOpt

mkYYY committed:
Add missing initialization for `setting`

mkYYY committed:
Fix compile errors

icchy committed:
improve optimizer store

icchy committed:
on-demand usage of SelectedCaseHistogram

icchy committed:
initialize optimizer keys for MOpt

icchy committed:
fix uninitialized bug for SelectedCaseHistogram

icchy committed:
use MOpt HierarFlow

icchy committed:
add test for MOpt

icchy committed:
fix mode transition

icchy committed:
fix initial value of core_mode, add ShowStats

icchy committed:
format

icchy committed:
rename havoc and splice for MOpt

icchy committed:
fix initial value for limit, pacemaker mode check

icchy committed:
add forkserv mode

icchy committed:
avoid ambigious parameter name *puppet*

icchy committed:
fix havoc_finds

icchy committed:
move store keys to AFL

- HavocFinds
- LastSpliceCycle

icchy committed:
use alias for MOptMutator

icchy committed:
explicitly call BuildFuzzFlow

icchy committed:
zerofill for SelectedCaseHistogram

icchy committed:
make mutop_optimizer shared_ptr

icchy committed:
merge MOptCaseDistrib into MOptOptimizer

icchy committed:
restore fuzz flow

icchy committed:
add license header, include guard

icchy committed:
fix typo

icchy committed:
add docs for MOpt

potetisensei commented:
> PR reviewers should review this PR in terms of the checklist below before moving on to a detailed code review. Please comment on their initial response by filling in the checklist below.

#### Optional Entries
- [x] The reviewer assigned more reviewers if needed.
- [x] The reviewer noted that it is necessary to break out some of the changes in this PR into other PRs if needed.
- [x] The reviewer noted that the initial response is insufficient if needed.

#### Mandatory Entries
- [x] The title of this PR summarizes the changes made by this PR properly.
- [x] The target branch of this PR is as intended.
- [x] The reviewer understands the issues in this PR.
- [x] The reviewer plans to review with an appropriate workload based on the importance of this PR.

Commit:
Fix optimizer/store.cpp fails to compile

Old nlohmann JSON including one provided by Ubuntu 20.04 has issue that
cannot exist with C++17 std::any in same translation unit.

As the result, MOpt implementation fails to comple on such environment
with logging enabled.

This change isolate the header into std::any related things and
logging(internally using nlohmann:json) related things.

Although the change fixed compilation failure avobe, other problem (lack of header file) is preventing whole build to success, and so, no any test is done.

icchy committed:
remove unused header

Commit:
Remove override on MOptFuzzer::BuildFuzzFlow()

potetisensei commented on algorithms/mopt/mopt_hierarflow_routines.cpp:
```diff
@@ -0,0 +1,213 @@
+  }
+
+  if (state.pacemaker_mode) {
+    // TODO
```
what kind of TODO?

icchy replied:
fixed in 2856b4dc4eff8de57f08c750de5945b8156dc17b
forgot to implement algorithm to exit pacemaker mode

potetisensei commented on algorithms/mopt/mopt_hierarflow_routines.cpp:
```diff
@@ -0,0 +1,213 @@
+
+  // update havoc_operator_finds
+  for (size_t i = 0; i < havoc_operator_finds.size(); i++) {
+    mopt->havoc_operator_finds[state.core_mode ? 1 : 0][i] +=
```
This may be a minute thing, but it is a bit unclear what the immediate values 0 and 1 are pointing out.
I would prefer something like `havoc_operator_finds[state.core_mode ? CORE_MODE : PILOT_MODE]` or  `havoc_operator_finds[state.fuzz_mode]` (by making state.core_mode `enum MOptMode`).

icchy replied:
fixed in 896319460a5f1cd0bb1c0934868b610193412ccf

potetisensei replied:
awesome

potetisensei replied:
Oh wait there seem some parts you forgot to fix.

https://github.com/fuzzuf/fuzzuf/pull/84/files#diff-77ad669e643808933880a044def0763a1788478a9f9a42bea615b130fdb42438R79-R87
https://github.com/fuzzuf/fuzzuf/pull/84/files#diff-77ad669e643808933880a044def0763a1788478a9f9a42bea615b130fdb42438R99-R102

potetisensei replied:
https://github.com/fuzzuf/fuzzuf/pull/84/files#diff-6a8c4972fbfb3703282f4f41c9b8a685401f29bc8c395d701a71bb15fdf0d496R58-R61

potetisensei replied:
https://github.com/fuzzuf/fuzzuf/pull/84/files#diff-6a8c4972fbfb3703282f4f41c9b8a685401f29bc8c395d701a71bb15fdf0d496R118

icchy replied:
fixed in 28bc3f78cad5b18ff8050e71fbad85363d630343

potetisensei replied:
Overlooking one:
https://github.com/fuzzuf/fuzzuf/pull/84/files#diff-6a8c4972fbfb3703282f4f41c9b8a685401f29bc8c395d701a71bb15fdf0d496R116

icchy replied:
I intentionally left this one as it doesn't have nothing to do with the type of MOptMode.

potetisensei replied:
Then why not use `NUM_CASE` ? Are those sometimes not the same?

icchy replied:
the point makes sense. applied in 48ca48478d07834d188c4590f0f692a0dec6aee2

potetisensei commented on algorithms/mopt/mopt_hierarflow_routines.cpp:
```diff
@@ -0,0 +1,213 @@
+      for (size_t i = 0; i < selected_case_histogram.size(); i++) {
+        double score = 0.0;
+        if (selected_case_histogram[i] > 0) {
+          score = (mopt->havoc_operator_finds[0][i] +
+                   mopt->havoc_operator_finds[1][i]) /
```
Here the same as
https://github.com/fuzzuf/fuzzuf/pull/84/files#r934244591

potetisensei commented on algorithms/mopt/mopt_hierarflow_routines.cpp:
```diff
@@ -0,0 +1,213 @@
+  auto selected_case_histogram = fuzzuf::optimizer::Store::GetInstance().Get(
+      fuzzuf::optimizer::keys::SelectedCaseHistogram, true);
+
+  // pilot mode (update local best)
+  if (!state.core_mode) {
+    if (unlikely(new_testcases > mopt::option::GetPeriodPilot<MOptTag>())) {
+      for (size_t i = 0; i < selected_case_histogram.size(); i++) {
+        double score = 0.0;
+        if (selected_case_histogram[i] > 0) {
+          score = (mopt->havoc_operator_finds[0][i] +
+                   mopt->havoc_operator_finds[1][i]) /
```
Here the same as
https://github.com/fuzzuf/fuzzuf/pull/84/files#r934244591

Moreover, it is not obvious for those who are not familiar with MOpt that, if the current mode is not "Core Mode", then it should be "Pilot Mode".
Although the top comment appropriately explains it, I think for example `state.mode == PILOT_MODE` would express it more directly

icchy replied:
fiexed in 896319460a5f1cd0bb1c0934868b610193412ccf

potetisensei commented on algorithms/mopt/mopt_optimizer.cpp:
```diff
@@ -0,0 +1,180 @@
+}
+
+void MOptOptimizer::UpdateGlobalBest() {
+  if (unlikely(time == 0)) {
+    // TODO?
+  }
```
what kind of TODO?

icchy replied:
removed as it is not required

potetisensei commented on algorithms/mopt/mopt_optimizer.cpp:
```diff
@@ -0,0 +1,180 @@
+void MOptOptimizer::UpdateLocalBest() {
+  auto& p = swarm[idx];
+
+  if (unlikely(time == 0)) {
+    for (size_t i = 0; i < p.fitness.size(); i++) {
+      p.best_fitness[i] = p.fitness[i];
+      p.best_position[i] = p.position[i];
+    }
+  }
```
I think, if you do this, the following loop should be included in `if (time ==0) {} else { ... }`.
But personally I prefer removing this block and initializing p.best_fitness[i] with -1 or -∞ or something

potetisensei replied:
Also, it seems `time` is incremented nowhere.

potetisensei replied:
Considering that `UpdateLocalBest` and `UpdateGlobalBest` both want to have initialization while their initialization takes place at different times, it would be dangerous to depend on the sole value `time`.
So I recommend you to remove the usage of `time`.

potetisensei commented on algorithms/mopt/mopt_optimizer.cpp:
```diff
@@ -0,0 +1,180 @@
+}
+
+void MOptOptimizer::UpdateInertia() {
+  w = (W_INIT - W_END) * (G_MAX - g_now) / G_MAX + W_END;
+  ++g_now %= G_MAX;
```
I think these values are a kind of hyperparameter and can be transformed into constexpr functions as done for AFL's hyperparameters.

icchy replied:
is it possible to convert this function to constexpr even it depends on non-const variable (g_now)?

potetisensei replied:
No I just meant that, for example, converting `W_INIT` into `constexpr SomeIntTypeIDontKnow GetWInit() { return SOME_CONSTANT_VALUE; }`.

icchy replied:
is it assumed to be implemented in MOptOption?

potetisensei replied:
I suppose so. But I'm doubting this conversion is needed considering that nothing will want to customize either PSO or MOpt

icchy replied:
so let's leave this one as it is :)

potetisensei commented on algorithms/mopt/mopt_optimizer.cpp:
```diff
@@ -0,0 +1,180 @@
+bool MOptOptimizer::IncrementSwarmIdx() {
+  idx++;
+  idx %= swarm.size();
+  return idx == 0;
```
The name `IncrementSwarmIdx` does not suggest that this function returns true if idx did a lap and becomes 0. But I don't have a good idea to resolve it. Renaming it to `IncrementSwarmIdxAndCheckIfMadeRound` might be slightly better...
But any ideas?

icchy replied:
fixed in baf38611e98a2c7d89aa9bf6995a83c684de32a5

rename `IncrementSwarmIdx` to `NextSwarmIdx` and made it return incremented index

potetisensei replied:
That's smart.
But I think you forgot to replace the type of return values.
https://github.com/fuzzuf/fuzzuf/commit/baf38611e98a2c7d89aa9bf6995a83c684de32a5#diff-6a8c4972fbfb3703282f4f41c9b8a685401f29bc8c395d701a71bb15fdf0d496R118

icchy replied:
fixed in 91e3ca33c40da44d60aa47847b59a2c85d5c1dd0

potetisensei commented on include/fuzzuf/algorithms/afl/templates/afl_state.hpp:
```diff
@@ -46,7 +46,7 @@ template <class Testcase>
     std::shared_ptr<const AFLSetting> setting,
     std::shared_ptr<executor::AFLExecutorInterface> executor,
-    std::unique_ptr<optimizer::Optimizer<u32>>&& _mutop_optimizer)
+    std::shared_ptr<optimizer::Optimizer<u32>>&& _mutop_optimizer)
```
I think there is no longer the reason to use rvalue reference now that we replaced unique_ptr with shared_ptr.

icchy replied:
fixed in 6b50cebeecd293f89664bc2a8165c3b89bc7ff35

potetisensei commented on include/fuzzuf/algorithms/afl/templates/afl_mutation_hierarflow_routines.hpp:
```diff
@@ -1,7 +1,7 @@
 /*
```
This file is only formatted, right?

icchy replied:
There are diffs by adding `havoc_operator_finds` and `last_splice_cycle`. Sorry for the commit with formatting and editing simultaneously.

```diff
diff --git a/old.hpp b/new.hpp
index 54eb619..90b0049 100644
--- a/old.hpp
+++ b/new.hpp
@@ -17,13 +17,12 @@
  */
 #pragma once
 
-#include "fuzzuf/exec_input/exec_input.hpp"
-#include "fuzzuf/utils/common.hpp"
-
 #include "fuzzuf/algorithms/afl/afl_mutator.hpp"
 #include "fuzzuf/algorithms/afl/afl_util.hpp"
-
+#include "fuzzuf/exec_input/exec_input.hpp"
+#include "fuzzuf/optimizer/keys.hpp"
 #include "fuzzuf/optimizer/optimizer.hpp"
+#include "fuzzuf/utils/common.hpp"
 
 namespace fuzzuf::algorithm::afl::routine::mutation {
 
@@ -35,7 +34,6 @@ BitFlip1WithAutoDictBuildTemplate<State>::BitFlip1WithAutoDictBuildTemplate(
 template <class State>
 AFLMutCalleeRef<State> BitFlip1WithAutoDictBuildTemplate<State>::operator()(
     AFLMutatorTemplate<State> &mutator) {
-
   /*********************************************
    * SIMPLE BITFLIP (+dictionary construction) *
    *********************************************/
@@ -166,7 +164,6 @@ AFLMutCalleeRef<State> ByteFlip1WithEffMapBuildTemplate<State>::operator()(
   if (state.eff_cnt != EFF_ALEN<Tag>(mutator.GetLen()) &&
       state.eff_cnt * 100 / EFF_ALEN<Tag>(mutator.GetLen()) >
           option::GetEffMaxPerc(state)) {
-
     std::memset(&state.eff_map[0], 1, EFF_ALEN<Tag>(mutator.GetLen()));
     state.blocks_eff_select += EFF_ALEN<Tag>(mutator.GetLen());
   } else {
@@ -404,6 +401,13 @@ bool HavocBaseTemplate<State>::DoHavoc(
 
   u64 havoc_queued = state.queued_paths;
 
+  std::array<u64, fuzzuf::mutator::NUM_CASE> havoc_operator_finds;
+  bool useHavocOperatorFinds =
+      fuzzuf::optimizer::Store::GetInstance().Exists(
+          fuzzuf::optimizer::keys::SelectedCaseHistogram) &&
+      fuzzuf::optimizer::Store::GetInstance().Exists(
+          fuzzuf::optimizer::keys::SelectedCaseHistogram);
+
   /* We essentially just do several thousand runs (depending on perf_score)
      where we take the input file and make random stacked tweaks. */
 
@@ -418,9 +422,27 @@ bool HavocBaseTemplate<State>::DoHavoc(
     mutator.Havoc(use_stacking, state.extras, state.a_extras, mutop_optimizer,
                   custom_cases);
 
+    u64 havoc_finds = state.queued_paths + state.unique_crashes;
+
     if (this->CallSuccessors(mutator.GetBuf(), mutator.GetLen()))
       return true;
 
+    havoc_finds = state.queued_paths + state.unique_crashes - havoc_finds;
+
+    if (unlikely(havoc_finds > 0) && useHavocOperatorFinds) {
+      auto selected_case_histogram =
+          fuzzuf::optimizer::Store::GetInstance().Get(
+              fuzzuf::optimizer::keys::SelectedCaseHistogram, true);
+      for (size_t i = 0; i < selected_case_histogram.size(); i++) {
+        if (selected_case_histogram[i] > 0) {
+          havoc_operator_finds[i] += havoc_finds;
+        }
+      }
+
+      fuzzuf::optimizer::Store::GetInstance().Set(
+          fuzzuf::optimizer::keys::HavocOperatorFinds, havoc_operator_finds);
+    }
+
     /* out_buf might have been mangled a bit, so let's restore it to its
        original size and shape. */
 
@@ -494,7 +516,6 @@ SplicingTemplate<State>::operator()(AFLMutatorTemplate<State> &mutator) {
   u32 splice_cycle = 0;
   while (splice_cycle++ < option::GetSpliceCycles(state) &&
          state.queued_paths > 1 && mutator.GetSource().GetLen() > 1) {
-
     /* Pick a random queue entry and seek to it. Don't splice with yourself. */
 
     u32 tid;
@@ -544,6 +565,8 @@ SplicingTemplate<State>::operator()(AFLMutatorTemplate<State> &mutator) {
 
     mutator.RestoreSplice();
   }
+  fuzzuf::optimizer::Store::GetInstance().Set(
+      fuzzuf::optimizer::keys::LastSpliceCycle, splice_cycle);
 
   return this->GoToDefaultNext();
 }
```

icchy replied:
close this as you opened other comments related to this file

potetisensei commented on include/fuzzuf/algorithms/afl/afl_state.hpp:
```diff
@@ -61,7 +61,7 @@ struct AFLStateTemplate {
       std::shared_ptr<const AFLSetting> setting,
       std::shared_ptr<executor::AFLExecutorInterface> executor,
-      std::unique_ptr<optimizer::Optimizer<u32>> &&mutop_optimizer);
+      std::shared_ptr<optimizer::Optimizer<u32>> &&mutop_optimizer);
```
The same as https://github.com/fuzzuf/fuzzuf/pull/84/files#r935112110

icchy replied:
fixed in 6b50cebeecd293f89664bc2a8165c3b89bc7ff35

potetisensei commented on docs/algorithms/mopt/implementation_en.md:
```diff
@@ -0,0 +1,16 @@
+
+## Differences from the original implementation
+
+Since original AFL-MOpt implements their method in a weird way, fuzzuf MOpt rewrote all algorithms to reproduce original MOpt-AFL behaviour with fuzzuf architecture. Basically there is no difference between MOpt-AFL and fuzzuf MOpt.
```
I think "weird" is not an appropriate word to describe the situation. It should be replaced with a more concrete and moderate word.

icchy replied:
would be fixed in another comment

potetisensei commented on include/fuzzuf/mutator/mutator.hpp:
```diff
@@ -306,13 +306,21 @@ void Mutator<Tag>::Havoc(u32 stacking, const std::vector<AFLDictData> &extras,
+
   for (std::size_t i = 0; i < stacking; i++) {
     u32 r = mutop_optimizer.CalcValue();
 
     switch (r) {
       case FLIP1:
         /* Flip a single bit somewhere. Spooky! */
         FlipBit(UR(len << 3), 1);
+        selected_case_histogram[r] += 1;
```
If you do `selected_case_histogram[r] += 1;` in every case, why not move it to the top of this selected area?

icchy replied:
It should not be incremented under some conditions and cases like `FLIP16`, `CLONE_BYTES` etc.
https://github.com/fuzzuf/fuzzuf/blob/144645f07ba1893335536a406347bfe505b5f62b/include/fuzzuf/mutator/mutator.hpp#L356
https://github.com/fuzzuf/fuzzuf/blob/144645f07ba1893335536a406347bfe505b5f62b/include/fuzzuf/mutator/mutator.hpp#L528-L551

This is from original MOpt-AFL
https://github.com/puppet-meteor/MOpt-AFL/blob/master/MOpt/afl-fuzz.c#L8351-L8396

but not sure if there is such a case actually not incrementing `selected_case_histogram`

potetisensei replied:
I got it. Then this seems legit. I'm now feeling switch-case itself is inconvenient.

potetisensei commented on include/fuzzuf/mutator/mutator.hpp:
```diff
@@ -654,8 +687,15 @@ void Mutator<Tag>::Havoc(u32 stacking, const std::vector<AFLDictData> &extras,
 
       default:
         custom_cases(r, outbuf, len, extras, a_extras);
+        selected_case_histogram[r] += 1;
```
In the default case, r can be greater than the size of `selected_case_histogram`, can't it?
Is it no problem?

icchy replied:
fixed in 621077e8c314ae8959f7057d19c82b3026068b60

potetisensei commented on include/fuzzuf/optimizer/pso.hpp:
```diff
@@ -0,0 +1,99 @@
+template <size_t Dimension, size_t ParticleNum>
+class PSO : public Optimizer<std::array<double, Dimension>> {
+ public:
+  PSO(double min_position, double max_position, double min_velocity,
+      double max_velocity, double w = 0.729, double c1 = 1.49445,
+      double c2 = 1.49445);
+  ~PSO();
+
+  void Init();
+  std::array<double, Dimension> GetCurParticle();
+  void SetScore(double);
+  std::array<double, Dimension> CalcValue() override;  // return global best
+  void UpdateLocalBest();
+  void UpdateGlobalBest();
+
+ protected:
+  void UpdatePositions();
+  void UpdateVelocities();
+
+  size_t idx = 0;
+  std::uint64_t time = 0;
+  std::array<Particle<Dimension>, ParticleNum> swarm;
+
+  // global best
+  std::array<double, Dimension> best_position;
+  double best_fitness;
+
+  // constraints
+  double min_position;
+  double max_position;
+  double min_velocity;
+  double max_velocity;
+
+  // parameters
+  double w = 0.729;     // inertia weight
+  double c1 = 1.49445;  // cognitive weight
+  double c2 = 1.49445;  // social weight
```
If you set the default values of w, c1, and c2 by the default arguments of the constructor, you don't have to write default initialization at the bottom. And vice versa.

icchy replied:
fixed in 4d52bbddca8588eaec69ce7f203438bb98e57dc5

potetisensei commented on include/fuzzuf/optimizer/pso.hpp:
```diff
@@ -0,0 +1,99 @@
+
+}  // namespace fuzzuf::optimizer
+
+#include "fuzzuf/optimizer/templates/pso.hpp"
```
If possible, I recommend you to merge `fuzzuf/optimizer/templates/pso.hpp` and `fuzzuf/optimizer/pso.hpp`.
Recall that we used this pattern to reduce git diffs. This pattern is now not mandatory.

icchy replied:
fixed in 83ec6c5338d5a3177a259af6a6a0250b53214163

potetisensei commented on include/fuzzuf/optimizer/store.hpp:
```diff
@@ -19,78 +19,108 @@
-    // may throw std::bad_any_cast
-    return std::any_cast<Type>(data[key.name]);
+template <typename Type>
+void Store::InitKey(const StoreKey<Type>& key) {
+  data[key.name] = std::any_cast<Type>(Type());
+}
+
+template <typename Type>
+void Store::InitKey(const StoreKey<Type>& key, Type val) {
+  data[key.name] = std::any_cast<Type>(Type(val));
 }
```
If we arrange that the value contained in Store should be initialized with `InitKey` first, then probably we can check if `InitKey` is called twice with the same key. This is intended to prevent two `StoreKey<T>` variables from wrongly having the same key string like `StoreKey<int> a {"duplicated"}, b {"duplicated"};`

icchy replied:
fixed in 164c51cd4bee3a09c180c9591d3cf6d93c2b53b3

potetisensei commented on include/fuzzuf/optimizer/pso.hpp:
```diff
@@ -0,0 +1,99 @@
+/*
```
Is this file still needed?

icchy replied:
This is not required for MOpt, just for reference implementation of original PSO as some parts of MOpt copied from here.

icchy replied:
`class Particle` is used in MOptOptimizer which is implemented in PSO.

potetisensei commented on include/fuzzuf/optimizer/templates/pso.hpp:
```diff
@@ -0,0 +1,154 @@
+/*
```
Is this file still needed?

icchy replied:
fixed in 83ec6c5338d5a3177a259af6a6a0250b53214163

potetisensei commented on include/fuzzuf/optimizer/store.hpp:
```diff
@@ -19,78 +19,108 @@
-    return std::any_cast<Type>(data[key.name]);
+template <typename Type>
+void Store::InitKey(const StoreKey<Type>& key) {
+  data[key.name] = std::any_cast<Type>(Type());
```
Is this any_cast needed? I think any_cast is used when we want to extract the actual value from std::any

icchy replied:
fixed in 4e88d3cae1b6142f179d9df22663aedbc31bd6de

potetisensei commented on include/fuzzuf/optimizer/store.hpp:
```diff
@@ -19,78 +19,108 @@
+
+template <typename Type>
+void Store::InitKey(const StoreKey<Type>& key, Type val) {
+  data[key.name] = std::any_cast<Type>(Type(val));
```
The same as https://github.com/fuzzuf/fuzzuf/pull/84/files#r935380292

potetisensei replied:
Moreover, I don't see any necessity of `Type(val)`. The argument `Type val` itself is copy-constructed from a passed value when we call this function. so there is no need for copy-construction again with `Type(val)`

icchy replied:
fixed in 164c51cd4bee3a09c180c9591d3cf6d93c2b53b3

potetisensei replied:
I think you are mentioning a wrong commit and in fact `Type(val)` is not fixed.
https://github.com/fuzzuf/fuzzuf/pull/84/files#diff-701a007040d1ab1542a058ba3761fc8531ed90ee0e4fa81ffedb95b44a1bf625R94

icchy replied:
fixed in 3c76ad015095b5e88fdbc63f83da4442da0d4c07

potetisensei commented on include/fuzzuf/cli/fuzzer/mopt/build_mopt_fuzzer_from_args.hpp:
```diff
@@ -0,0 +1,243 @@
+namespace po = boost::program_options;
+
+struct MOptFuzzerOptions {
+  bool forksrv;                        // Optional
+  std::vector<std::string> dict_file;  // Optional
+  bool frida_mode;                     // Optional
+  u64 mopt_limit_time = 1;
+  u64 mopt_most_time = 0;
+
+  // Default values
+  MOptFuzzerOptions() : forksrv(true), frida_mode(false){};
```
I would like this part to be unified with default initialization or without it.

icchy replied:
fixed in 0a7be14649ee0df3dda4605efe02a2b03918f748

potetisensei commented on algorithms/afl/afl_havoc_case_distrib.cpp:
```diff
@@ -0,0 +1,79 @@
+
+    bool has_extras  = !extras.empty();
+    bool has_aextras = !a_extras.empty();
+    return static_cast<HavocCase>(dists[has_extras][has_aextras]());
```
I think casting a return value to HavocCase is not necessary(because the type of return value is u32)

potetisensei replied:
Oh this part seems irrelevant to this PR

icchy committed:
make core_mode enum MOptMode

icchy committed:
fix selected_case_hitogram oob in default case

icchy committed:
merge templates into declaration

potetisensei commented on algorithms/mopt/mopt_hierarflow_routines.cpp:
```diff
@@ -0,0 +1,213 @@
+        double score = 0.0;
+        if (selected_case_histogram[i] > 0) {
+          score = (mopt->havoc_operator_finds[0][i] +
+                   mopt->havoc_operator_finds[1][i]) /
```
Here, you don't need to add `havoc_operator_finds` of core mode.

icchy replied:
fixed in b386bc7a39218cd286ab99b8e2fc51af97ead618

potetisensei commented on algorithms/mopt/mopt_hierarflow_routines.cpp:
```diff
@@ -0,0 +1,213 @@
+        mopt->SetScore(i, score);
+      }
+      mopt->UpdateLocalBest();
+
```
After (trying) the update of local best, you need to reset `havoc_operator_finds` of pilot mode and the corresponding `(accm_)selected_case_histogram` .

icchy replied:
fixed in b386bc7a39218cd286ab99b8e2fc51af97ead618

potetisensei replied:
Oh wait. There is another issue.

You don't check if the new score is higher than the one recorded in the optimizer.

Compare the followings:
https://github.com/puppet-meteor/MOpt-AFL/blob/master/MOpt/afl-fuzz.c#L8717-L8721

and

https://github.com/fuzzuf/fuzzuf/pull/84/files#diff-77ad669e643808933880a044def0763a1788478a9f9a42bea615b130fdb42438R83
https://github.com/fuzzuf/fuzzuf/pull/84/files#diff-6a8c4972fbfb3703282f4f41c9b8a685401f29bc8c395d701a71bb15fdf0d496R86-R89

icchy replied:
`UpdateLocalBest` will check if the score is higher or not, which is called after all scores set
So I think this would not be an issue.

potetisensei replied:
Ah that's right 
sorry don't care

potetisensei commented on algorithms/mopt/mopt_hierarflow_routines.cpp:
```diff
@@ -0,0 +1,213 @@
+  if (state.core_mode) {
+    if (unlikely(new_testcases > mopt::option::GetPeriodCore<MOptTag>())) {
+      mopt->UpdateGlobalBest();
+      state.core_mode = false;
```
Seems you need to reset `havoc_operator_finds` of core mode here.
https://github.com/puppet-meteor/MOpt-AFL/blob/master/MOpt/afl-fuzz.c#L10601-L10607

icchy replied:
fixed in b386bc7a39218cd286ab99b8e2fc51af97ead618

potetisensei commented on algorithms/mopt/mopt_hierarflow_routines.cpp:
```diff
@@ -0,0 +1,213 @@
+        if (selected_case_histogram[i] > 0) {
+          score = (mopt->havoc_operator_finds[0][i] +
+                   mopt->havoc_operator_finds[1][i]) /
+                  selected_case_histogram[i];
```
You have to add `(accm_)selected_case_histogram` as a member of MOptOptimizer, like `mopt->(accm_)havoc_operator_finds`

icchy replied:
fixed in b386bc7a39218cd286ab99b8e2fc51af97ead618

potetisensei commented on algorithms/mopt/mopt_optimizer.cpp:
```diff
@@ -0,0 +1,180 @@
+
+void MOptOptimizer::UpdateInertia() {
+  w = (W_INIT - W_END) * (G_MAX - g_now) / G_MAX + W_END;
+  ++g_now %= G_MAX;
```
Don't you have to do `++g_now %= G_MAX;` before `w = (W_INIT - W_END) * ...` ?
https://github.com/puppet-meteor/MOpt-AFL/blob/master/MOpt/afl-fuzz.c#L10625

icchy replied:
fixed in d23ed32b16a3d3a158ee1ab6114485a293263e8b

potetisensei commented on algorithms/mopt/mopt_optimizer.cpp:
```diff
@@ -0,0 +1,180 @@
+  bool no_extras = extras.empty();
+  bool no_aextras = a_extras.empty();
+
+  auto weights(swarm[idx].best_position);
```
Here, probably `position` is correct instead of `best_position`.

icchy replied:
fixed in 520372bb926882100cd863f458fb878d5249d833

potetisensei commented on algorithms/mopt/mopt_hierarflow_routines.cpp:
```diff
@@ -0,0 +1,213 @@
+      mopt->UpdateLocalBest();
+
+      if (mopt->IncrementSwarmIdx()) {  // all swarms are visited
+        state.core_mode = true;
```
Here, you need to find out the index of the best swarm and remember it somehow to use it in core mode.
https://github.com/puppet-meteor/MOpt-AFL/blob/master/MOpt/afl-fuzz.c#L8740-L8748

icchy replied:
fixed in 520372bb926882100cd863f458fb878d5249d833

potetisensei commented on algorithms/mopt/mopt_optimizer.cpp:
```diff
@@ -0,0 +1,180 @@
+  bool no_extras = extras.empty();
+  bool no_aextras = a_extras.empty();
+
+  auto weights(swarm[idx].best_position);
```
Here, you need to switch `idx` and `best_idx` properly according to the current mode. First you have to add `best_idx`.

icchy replied:
fixed in 520372bb926882100cd863f458fb878d5249d833

potetisensei commented on docs/algorithms/mopt/algorithm_en.md:
```diff
@@ -0,0 +1,43 @@
+
+## What is MOpt
+
+[MOpt-AFL](https://github.com/puppet-meteor/MOpt-AFL) is an example implementation for MOpt[^usenix19] paper applying customized PSO (Particle Swarm Optimization) algorithm to AFL mutation scheduling.
```
I think their algorithm is not even a *customized* PSO, but original heuristics inspired by PSO.
This is because the original PSO and MOpt's PSO-like algorithms don't share the problem setting and the objective function at all. 
This can be emphasized to some extent considering that I saw some people misunderstand this point.

icchy replied:
fixed in 35e99df42d8ec0956375928d3a4b0ff642d76cef

potetisensei reviewed (changes requested):
I read through the changes introduced by this PR, and commented all the points that would need fixing. 
I don't see any problems in the goal and general design of this PR, so I can approve this when the comments are all cleared.

icchy committed:
fix initialization for best_fitness/positions

icchy committed:
no rvalue reference is required for shared_ptr

icchy committed:
raise ERROR on duplicate store key

icchy committed:
fix compile error on C++17

potetisensei commented on include/fuzzuf/algorithms/afl/templates/afl_mutation_hierarflow_routines.hpp:
```diff
@@ -17,536 +17,549 @@
 
-    for (state.stage_cur = 0; state.stage_cur < state.stage_max; state.stage_cur++) {
-        using afl::util::UR;
+  std::array<u64, fuzzuf::mutator::NUM_CASE> havoc_operator_finds;
```
I think this is not initialized and hence causes Use of Uninitialized Value.

potetisensei replied:
Maybe you should check the other files where std::array is used (e.g. MOptParticle::fitness).

icchy replied:
fixed in 3aed518b55a335259bd324e710f4d0bac862ddef

icchy replied:
and fixed uninitialized things in 17e719804bfb4cf61b175ef093e608febae6b240

potetisensei commented on include/fuzzuf/algorithms/afl/templates/afl_mutation_hierarflow_routines.hpp:
```diff
@@ -17,536 +17,549 @@
+    if (this->CallSuccessors(mutator.GetBuf(), mutator.GetLen())) return true;
 
-        if (state.queued_paths != havoc_queued) {
-            if (perf_score <= option::GetHavocMaxMult(state) * 100) {
-                state.stage_max *= 2;
-                perf_score *= 2;
-            }
+    havoc_finds = state.queued_paths + state.unique_crashes - havoc_finds;
```
I don't like to see the same variable used different ways. 
I prefer something like, `u64 prev_havoc_finds = state.queued_paths + state.unique_crashes;` and `havoc_finds_diff = state.queued_paths + state.unique_crashes - prev_havoc_finds;`.

icchy replied:
fixed in 3aed518b55a335259bd324e710f4d0bac862ddef

icchy committed:
make w, c1, c2 as default params of constructor

icchy committed:
remove unnecessary std::any_cast

icchy committed:
initialize mopt_{limit,most}_time properly

icchy committed:
initialize havoc_oeprator_finds, rename vars

icchy committed:
initialize std::array

icchy committed:
rename IncrementSwarmIdx -> NextSwarmIdx

icchy committed:
fix order of ++g_now

icchy committed:
fix algorithm to use accumulated ones

icchy committed:
make mode/pacemaker_mode MOptOptimizer member

icchy committed:
add best_idx, last_havoc_finds for swarm_fitness

icchy committed:
embedded inertia params

icchy committed:
fix pacemaker algorithm

icchy committed:
fix docs

potetisensei commented on docs/algorithms/mopt/algorithm_en.md:
```diff
@@ -0,0 +1,46 @@
+## What is MOpt
+
+[MOpt-AFL](https://github.com/puppet-meteor/MOpt-AFL) is an example implementation for MOpt[^usenix19] paper.
+We think their algorithm is not even a customized PSO, but original heuristics inspired by PSO.
+This is because the original PSO and MOpt's PSO-like algorithms don't share the problem setting and the objective function at all.
+This can be emphasized to some extent considering that I saw some people misunderstand this point.
+They optimize the selection of mutation operators to find more interesting inputs by generating probability distribution from particles optimized with PSO-like algorithm.
```
This part sounds a bit too aggressive and assertive to me for some reason. 
Moreover, if we say "we think", it seems like we are making some claims or clarifying our standpoint, which a third party cannot tell true or not. But, as a matter of fact, as I explained, this is true, so I would suggest the following so that the explanation sounds more objective and concise:  

```suggestion
It tries to optimize the selection of mutation operators to find more interesting inputs by generating probability distributions from *particles* calculated with its original heuristics inspired by PSO (Particle Swarm Optimization).
```

icchy replied:
thanks for great suggestion. reflected in 3f78ce24c2055672e02e9fa58ca2c19b4b453345

potetisensei commented on algorithms/mopt/mopt_hierarflow_routines.cpp:
```diff
@@ -0,0 +1,243 @@
+
+      if (mopt->NextSwarmIdx() == 0) {  // all swarms are visited
+        mopt->UpdateBestSwarmIdx();
+        mopt->mode = MOptMode::CoreMode;
```
Sorry for overlooking this, but don't we have to add the code corresponding to this?

https://github.com/puppet-meteor/MOpt-AFL/blob/master/MOpt/afl-fuzz.c#L8732-L8737

icchy replied:
you're right. fixed in 549e28888267fc779b932b99ac3279199635b3af

potetisensei reviewed (changes requested):
Only a few issues are left.

icchy committed:
change return value type of NextSwarmIdx

icchy committed:
not to call copy constructor on storing key

icchy committed:
fix docs expression

icchy committed:
reset counter when all swarms are visited

icchy committed:
make it clear which mode is specified

potetisensei commented on include/fuzzuf/optimizer/pso.hpp:
```diff
@@ -0,0 +1,229 @@
+  double best_fitness;
+};
+
+template <size_t Dimension, size_t ParticleNum>
```
If you will retain this file, then I have to review this class and request changes.
But the changes would be somewhat not small.
So considering the class is (and almost certainly will be) not used,  you can just comment something like "FIXME: bababa..." instead of applying the following changes if you don't want to take time.

icchy replied:
I believe all issues are resolved in b6207e50d21b625fb526c646059baf36fe5a43c4 (and I think this class would not be used as well)

potetisensei commented on include/fuzzuf/optimizer/pso.hpp:
```diff
@@ -0,0 +1,229 @@
+template <size_t Dimension, size_t ParticleNum>
+PSO<Dimension, ParticleNum>::~PSO() {}
+
+template <size_t Dimension, size_t ParticleNum>
+void PSO<Dimension, ParticleNum>::Init() {
+  for (auto& p : swarm) {
+    for (auto& pos : p.position)
+      pos = fuzzuf::utils::random::Random<double>(min_position, max_position);
+    p.velocity.fill(0);
+  }
+
+  idx = 0;
+  time = 0;
+}
```
Why not just do this in the constructor?

icchy replied:
fixed in b6207e50d21b625fb526c646059baf36fe5a43c4

potetisensei commented on include/fuzzuf/optimizer/pso.hpp:
```diff
@@ -0,0 +1,229 @@
+    : min_position(min_position),
+      max_position(max_position),
+      min_velocity(min_velocity),
+      max_velocity(max_velocity),
```
PSO::best_position is not initialized

icchy replied:
fixed in b6207e50d21b625fb526c646059baf36fe5a43c4

potetisensei commented on include/fuzzuf/optimizer/pso.hpp:
```diff
@@ -0,0 +1,229 @@
+void PSO<Dimension, ParticleNum>::UpdateLocalBest() {
+  auto& p = swarm[idx];
+
+  if (unlikely(time == 0)) {
+    p.best_position = p.position;
+    p.best_fitness = p.fitness;
+    return;
+  }
+
+  if (p.fitness < p.best_fitness ^
+      !opt_minimize) {  // not when optimize to maximize
+    p.best_position = p.position;
+    p.best_fitness = p.fitness;
+  }
+}
+
+template <size_t Dimension, size_t ParticleNum>
+void PSO<Dimension, ParticleNum>::UpdateGlobalBest() {
+  if (unlikely(time == 0)) {
+    best_fitness = swarm[0].best_fitness;
+    best_position = swarm[0].best_position;
+  }
```
As I said in MOptOptimizer, I don't like to see `time` is used here.
I think it would be better to remove `time` and initialize these values with some sentinels instead so that `UpdateLocalBest` and `UpdateGlobalBest` can be safely called any time in any order (e.g. what if users expect that `UpdateGlobalBest` works correctly even when `time == 0 && idx == 2 && swarm.size() == 4`).

icchy replied:
fixed in b6207e50d21b625fb526c646059baf36fe5a43c4

icchy committed:
some fix applied in pso.hpp

icchy committed:
use NUM_CASE instead of size of array

potetisensei reviewed (approved):
Great job
